pkg/cloud/baremetal/actuators/machine/actuator.go

```diff
@@ -498,6 +503,19 @@ func (a *Actuator) ensureAnnotation(ctx context.Context, machine *machinev1beta1
 	return true, a.client.Update(ctx, machine)
 }
 
+func (a *Actuator) ensureProviderID(ctx context.Context, machine *machinev1beta1.Machine, host *bmh.BareMetalHost) (error) {
+	if host.Status.Provisioning.ID != "" {
+		providerID := "baremetal:" + host.Status.Provisioning.ID
+		existingProviderID := machine.Spec.ProviderID
+		if existingProviderID == nil || *existingProviderID != providerID {
+			log.Printf("Setting ProviderID %s for machine %s.", providerID, machine.Name)
+			machine.Spec.ProviderID = &providerID
+			return a.client.Update(ctx, machine)
+		}
+	}
+	return nil
```

Comment:
should this return something to indicate that the function wasn't able to set the provider ID?

Reply:
I was expecting this to be called repeatedly and initially the BMH won't have any host.Status.Provisioning.ID, so it's reasonable to just return nil and do nothing until that's set?

Reply:
It would only be called repeatedly if there was more reconciliation to do.

I think the suggestion from @mhrivnak in slack to use the host resource name will eliminate the concern with not having a value, so the point is probably moot.

Reply:
Yeah so if we change this to e.g `baremetalhost:///namespace/name` that resolves the timing issue inherent in using the ironic ID - however we'd still have to ensure reconciliation continues until the machine.Status.NodeRef gets populated (at which point we can then populate the ProviderID on the node.

Reply:
> Yeah so if we change this to e.g `baremetalhost:///namespace/name` that resolves the timing issue inherent in using the ironic ID - however we'd still have to ensure reconciliation continues until the machine.Status.NodeRef gets populated (at which point we can then populate the ProviderID on the node.

What sets the NodeRef (I don't see it happening in this repo)? If it's another controller, that will trigger reconciliation of the Machine, running Update() here again, so we should be covered.

Reply:
I think it's set by the nodelink controller in the MAO repo, when the node becomes ready:

https://github.com/openshift/machine-api-operator/blob/master/pkg/controller/nodelink/nodelink_controller.go#L257

I don't see any platform specific logic in that controller though, so perhaps it's best to add the node update to the actuator, triggered by that event?